### PR TITLE
[ROCm][CI] Add attention-sink support to ROCm AITER sparse MLA

### PR DESCRIPTION
**Pull request by AndreasKaratzas**
HY-V4 introduced learnable attention sinks in [#54160](https://github.com/vllm-project/vllm/pull/54160), but its ROCm initialization was unsupported in [AMD build 12635](https://buildkite.com/vllm/amd-ci/builds/12635/list?jid=01a06dbb-c994-4dda-9277-8d85dae8eecb&tab=output) and [build 12653](https://buildkite.com/vllm/amd-ci/builds/12653/list?jid=01a070cd-0530-4d60-a5fd-a77dcc439bad&tab=output). This PR supplies reusable sparse-MLA sink support required by the model enablement in [#54405](https://github.com/vllm-project/vllm/pull/54405). Local MI300 validation also exposed empty-row NaNs, rejected FP16 inputs and missing native BF16 kernels for the 64-head bucket; those cases now execute correctly.

- Add a graph-visible, nonpersistent AITER operation returning LSE and apply virtual zero-value sinks with `logaddexp` and output rescaling; empty rows contribute exactly zero.
- Use the existing Triton fallback for FP16 and unsupported native head shapes, preserving query dtype and requested value width.
- Preserve ordinary sparse-MLA paths, validate sink metadata, reject unsupported DCP combinations and advertise the supported contiguous cache layouts.
- Cover ragged/dense inputs, FP8 scales, head padding, empty rows, metadata and operator contracts; use FP64 references and precision-based error budgets on CDNA 3/4.

Prepared with AI assistance (OpenAI Codex).

### CODE REVIEW

**Comment by AndreasKaratzas** [comment]
Tested stack: PR #54404 `68628f9b7a73` + PR #54405 `29c7fa202770`, model revision `4215ec29de87`, on 8x MI300X with TP=8.

| Check | Result | Evidence |
|---|---|---|
| Sink/LSE numerical reference | PASS | 18 passed, 2 expected gfx942 H64 skips; the exact 8-local-head BF16 path matched PyTorch. |
| 2,048-token sparse stress | PASS | Max output error `0.0018023`; max LSE error `9.54e-7`; finite and bitwise repeatable. |
| Supporting correctness suites | PASS | 137 additional targeted tests passed: decode, padding, schemas, metadata, MTP routing, registry, and selector. |
| Baseline full-model E2E | PASS | Sparse prefill/decode, ragged batch, graph replay, streaming, and tool parsing succeeded. |
| MTP full-model E2E | PASS | 3 speculative steps; 9/9 drafted tokens accepted; health remained HTTP 200. |
| Four-item quality probes | INFO | `no_think`: 2/4; `high`: 2 correct + 2 truncated at 128 tokens; GSM8K smoke: 3/4. These samples are not correctness benchmarks. |
| Residual scope | LIMIT | No CUDA/FlashMLA logit-parity run and no gfx950 or FP8-KV full-model E2E. |

So this impl looks functionally and numerically correct for the tested MI300X TP=8 configuration.

**Comment by indivats** [comment]
Independent validation on 8x **MI350X** (gfx950), TP=8 on #54594's head 4be7ce114, which contains this PR's current head 68628f9b7a73 and #54405's 29c7fa2027 unchanged. Model revision 4215ec29de87, base image vllm/vllm-openai-rocm@sha256:d53c0dd4:

- Sink/LSE + op registration on gfx950: 20 passed, 2 skipped (the same gfx942-specific H64 head-count skips).
- Focused suite: 162 passed, 2 skipped, 0 failed in 126.75s (superset of #54594's 142: it adds the two sink/op-registration files from this PR)
- Backend selection: `Using ROCM_AITER_MLA_SPARSE backend out of potential backends: ['ROCM_AITER_MLA_SPARSE']`.
- Sink applied: `HYV4 learnable sink enabled: forcing sparse MQA for prefill too, ...`.
- Full-model E2E, TP=8: served to `Application startup complete`; graph mode (enforce_eager=False, FULL_AND_PIECEWISE); KV 1,652,144 tokens / 144.53 GiB per GPU at --max-model-len 32768. serving requires VLLM_ROCM_USE_AITER=1 at process start.
- Functional ladder 5/5: /v1/models, chat completion, reasoning parser, no_think off-switch, tool calling (finish_reason=tool_calls).
- Reasoning under load: across all 1,319 GSM8K requests, 1,319/1,319 returned a non-empty `reasoning` channel and non-empty content (zero parser failures) and 95.34% of generated tokens were reasoning tokens (2,343,524 / 2,458,156). Median 481, p95 9,776, max 25,263. 
- Native MTP: draft model loaded (35 params); at k=3, mean accepted length 2.18 at concurrency 1 and 2.10 at concurrency 32.
- Full GSM8K (with explicit answer-format prompt): all 1,319 samples, temperature 0.9 / top_p 1.0 / seed 0, four request threads: 1,272/1,319 = 96.44%, 0 length truncations, 0 missing predictions, 0 request errors, 364.94 output tok/s over 112 minutes.

**Comment by dllehr-amd** [comment]
So a funny thing looking through this.  I can't find an existing use case where we checked `supports_dcp`   So this PR may have the unintended consequence of catching an attention backend that was supposed to be failing but wasn't if they had `supports_dcp = false` previously.  Dont suppose this is a problem per se.  Probably closing the gap for us

**Review by dllehr-amd — APPROVED** [review]
This one looks good.  @njhill  if I may pick your brain, the only caveat I see here is that we haven't looked for dcp support explicitly before, but now we will in backend.py.   This is probably the right thing to do?  But may escape something that maybe shouldn't have worked before.  You cool with it?

File: vllm/_aiter_ops.py [hunk 1]
```diff
@@ -624,6 +624,110 @@ def _rocm_aiter_mla_decode_fwd_impl(
     )
 
 
+def _rocm_aiter_mla_decode_fwd_lse_impl(
```
Comment (tjtanaa) [inline vllm/_aiter_ops.py:627]: We should refactor/extract out the common logic of these two functions and only create a wrapper that register them into different function signature (inplace and outofplace)
`_rocm_aiter_mla_decode_fwd_impl`. `_rocm_aiter_mla_decode_fwd_lse_impl`

Comment (AndreasKaratzas) [reply vllm/_aiter_ops.py:627]: Yep. We can do that in follow up along with other cleanup ops, for now since CI has already run, probably best to merge otherwise there ll soon be another merge conflict :sweat_smile:

File: vllm/v1/attention/backend.py [hunk 2]
```diff
@@ -226,6 +226,13 @@ def supports_pcp(cls) -> bool:
         except NotImplementedError:
             return False
 
+    @classmethod
+    def supports_dcp(cls) -> bool:
```
Comment (tjtanaa) [inline vllm/v1/attention/backend.py:230]: This could cause issue to all other backends. There are other backends that supports dcp e.g. Trion MLA backend. I am not familiar with which backend actually support dcp.

Comment (tjtanaa) [reply vllm/v1/attention/backend.py:230]: @AndreasKaratzas

Comment (AndreasKaratzas) [reply vllm/v1/attention/backend.py:230]: @tjtanaa Looking into it. Might push soon a follow-up for this.

Comment (AndreasKaratzas) [reply vllm/v1/attention/backend.py:230]: I checked the inheritance and the new method reads `get_impl_cls().supports_dcp`, and that implementation flag already defaulted to True before this PR. Triton MLA therefore remains eligible. However, I’m already preparing a follow-up that defaults the base to False and explicitly enables the existing DCP implementations, including Triton MLA, dense AITER MLA, and FlashAttention, with regression tests covering selection. Triton MLA DCP generation was verified locally on ROCm btw.
